File: clients/jvm/Squeue.java
// Minimal SqueueLite client for the JVM (Java 16+) — dependency-free, JSON-RPC 2.0.
//
// Uses java.nio Unix Domain Socket support (Java 16+). Kotlin / Scala / Clojure
// can call it directly. SqueueLite is *write-only*; read the SQLite file
// directly (read-only, WAL).
//
// Methods return the raw JSON-RPC 2.0 response line as a String so you can
// parse it with whatever JSON library you already use (Jackson, Gson, …) — or
// none. Successful responses contain a "result" key; error responses contain an
// "error" key with "code" and "message".
//
// Demo:  java clients/jvm/Squeue.java ./squeuelite.sock

package squeue;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.StandardProtocolFamily;
import java.net.UnixDomainSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

public final class Squeue implements AutoCloseable {
    private final String actorId;
    private final SocketChannel ch;
    private final AtomicLong counter = new AtomicLong(0);

    /**
     * Wrap binary data for a BLOB column parameter ({@code {"$blob": "<base64>"}}).
     * Pass an instance as a param value:
     * {@code db.execute("INSERT INTO files(data) VALUES (?)", List.of(new Squeue.Blob(bytes)))}
     */
    public record Blob(byte[] data) {}

    public Squeue(String socketPath, String actorId) throws IOException {
        this.actorId = actorId;
        this.ch = SocketChannel.open(StandardProtocolFamily.UNIX);
        this.ch.connect(UnixDomainSocketAddress.of(socketPath));
    }

    /** Run one statement as a single transaction. Returns the raw JSON-RPC 2.0 response line. */
    public String execute(String sql, List<Object> params) throws IOException {
        return execute(sql, params, null, null);
    }

    /** Run one statement, optionally with idempotency_key / run_id. */
    public String execute(String sql, List<Object> params, String idempotencyKey, String runId)
            throws IOException {
        String ops = "[{\"sql\":" + jsonStr(sql) + ",\"params\":" + jsonArray(params) + "}]";
        return sendExecute(ops, idempotencyKey, runId);
    }

    /** Admin method (stats / health / checkpoint). Returns the raw JSON-RPC 2.0 response line. */
    public String stats() throws IOException {
        return adminMethod("stats");
    }

    /** Admin method: health check. Returns the raw JSON-RPC 2.0 response line. */
    public String health() throws IOException {
        return adminMethod("health");
    }

    /** Admin method: WAL checkpoint. Returns the raw JSON-RPC 2.0 response line. */
    public String checkpoint() throws IOException {
        return adminMethod("checkpoint");
    }

    private String adminMethod(String method) throws IOException {
        long id = counter.incrementAndGet();
        writeLine("{\"jsonrpc\":\"2.0\",\"id\":" + id + ",\"method\":" + jsonStr(method) + "}");
        return readLine();
    }

    private String sendExecute(String operationsJson, String idempotencyKey, String runId)
            throws IOException {
        long id = counter.incrementAndGet();
        StringBuilder params = new StringBuilder();
        params.append("{\"actor_id\":").append(jsonStr(actorId));
        if (runId != null) params.append(",\"run_id\":").append(jsonStr(runId));
        if (idempotencyKey != null) params.append(",\"idempotency_key\":").append(jsonStr(idempotencyKey));
        params.append(",\"operations\":").append(operationsJson).append("}");

        StringBuilder b = new StringBuilder();
        b.append("{\"jsonrpc\":\"2.0\",\"id\":").append(id);
        b.append(",\"method\":\"execute\"");
        b.append(",\"params\":").append(params).append("}");
        writeLine(b.toString());
        return readLine();
    }

    private void writeLine(String json) throws IOException {
        ByteBuffer out = ByteBuffer.wrap((json + "\n").getBytes(StandardCharsets.UTF_8));
        while (out.hasRemaining()) {
            ch.write(out);
        }
    }

    private String readLine() throws IOException {
        ByteArrayOutputStream line = new ByteArrayOutputStream();
        ByteBuffer one = ByteBuffer.allocate(1);
        while (true) {
            one.clear();
            int n = ch.read(one);
            if (n < 0) throw new IOException("gateway closed the connection");
            if (n == 0) continue;
            byte b = one.get(0);
            if (b == '\n') break;
            line.write(b);
        }
        return line.toString(StandardCharsets.UTF_8);
    }

    // -- tiny JSON encoder (strings + scalar params only) ---------------------

    private static String jsonArray(List<Object> params) {
        StringBuilder b = new StringBuilder("[");
        for (int i = 0; i < params.size(); i++) {
            if (i > 0) b.append(",");
            b.append(jsonValue(params.get(i)));
        }
        return b.append("]").toString();
    }

    private static String jsonValue(Object v) {
        if (v == null) return "null";
        if (v instanceof Number || v instanceof Boolean) return v.toString();
        if (v instanceof Blob bl) {
            return "{\"$blob\":" + jsonStr(Base64.getEncoder().encodeToString(bl.data())) + "}";
        }
        return jsonStr(v.toString());
    }

    private static String jsonStr(String s) {
        StringBuilder b = new StringBuilder("\"");
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  b.append("\\\""); break;
                case '\\': b.append("\\\\"); break;
                case '\n': b.append("\\n");  break;
                case '\r': b.append("\\r");  break;
                case '\t': b.append("\\t");  break;
                default:
                    if (c < 0x20) b.append(String.format("\\u%04x", (int) c));
                    else b.append(c);
            }
        }
        return b.append("\"").toString();
    }

    @Override
    public void close() throws IOException {
        ch.close();
    }

    public static void main(String[] args) throws IOException {
        String path = args.length > 0 ? args[0] : "./squeuelite.sock";
        try (Squeue db = new Squeue(path, "demo-jvm")) {
            System.out.println("health: " + db.health());
            System.out.println("write : " + db.execute(
                    "INSERT INTO events(agent_id, kind) VALUES (?, ?)",
                    List.of("demo-jvm", "started"), "demo:1", null));
            System.out.println("stats : " + db.stats());
        }
    }
}

File: clients/deno/squeue.ts
// Minimal SqueueLite client for Deno — Unix Domain Socket, JSON-RPC 2.0.
//
// SqueueLite is *write-only*; read the SQLite file directly (read-only, WAL).
//
// Run the demo. A Unix-socket connect needs read+write permission on the socket
// file AND net access to the unix address:
//   deno run --allow-read --allow-write --allow-net clients/deno/squeue.ts ./squeuelite.sock
//
// Example:
//   import { Squeue } from "./squeue.ts";
//   const db = await Squeue.connect("/run/app/squeuelite.sock", "agent-deno");
//   const r = await db.execute(
//     "INSERT INTO events(agent_id, kind) VALUES (?, ?)",
//     ["agent-deno", "started"],
//     { idempotencyKey: "run-7:step-1", runId: "run-7" });
//   // r == { jsonrpc: "2.0", id: 1, result: { status: "committed", commit_seq: 12 } }
//   // Check r.result for success, r.error for failure.
//   db.close();

// Wrap binary data for a BLOB column parameter: {"$blob": "<base64>"}.
//   db.execute("INSERT INTO files(data) VALUES (?)", [blob(bytes)])
export function blob(data: Uint8Array): { $blob: string } {
  let bin = "";
  for (const b of data) bin += String.fromCharCode(b);
  return { $blob: btoa(bin) };
}

export interface WriteOpts {
  idempotencyKey?: string;
  runId?: string;
}

export class Squeue {
  #conn: Deno.UnixConn;
  #buf = "";
  #enc = new TextEncoder();
  #dec = new TextDecoder();
  #counter = 0;

  private constructor(conn: Deno.UnixConn, public actorId: string) {
    this.#conn = conn;
  }

  static async connect(path: string, actorId: string): Promise<Squeue> {
    const conn = await Deno.connect({ transport: "unix", path });
    return new Squeue(conn, actorId);
  }

  #nextId(): number {
    return ++this.#counter;
  }

  // Run one statement as a single atomic transaction.
  execute(sql: string, params: unknown[] = [], opts: WriteOpts = {}) {
    return this.transaction([[sql, params]], opts);
  }

  // Run several [sql, params] ops as ONE transaction (all-or-nothing).
  transaction(
    ops: Array<[string, unknown[]]>,
    { idempotencyKey, runId }: WriteOpts = {},
  ) {
    const rpcParams: Record<string, unknown> = {
      actor_id: this.actorId,
      operations: ops.map(([sql, params]) => ({ sql, params })),
    };
    if (runId != null) rpcParams.run_id = runId;
    if (idempotencyKey != null) rpcParams.idempotency_key = idempotencyKey;
    return this.#roundtrip({
      jsonrpc: "2.0",
      id: this.#nextId(),
      method: "execute",
      params: rpcParams,
    });
  }

  stats() {
    return this.#roundtrip({ jsonrpc: "2.0", id: this.#nextId(), method: "stats" });
  }
  health() {
    return this.#roundtrip({ jsonrpc: "2.0", id: this.#nextId(), method: "health" });
  }
  checkpoint() {
    return this.#roundtrip({ jsonrpc: "2.0", id: this.#nextId(), method: "checkpoint" });
  }

  async #roundtrip(obj: unknown): Promise<Record<string, unknown>> {
    await this.#conn.write(this.#enc.encode(JSON.stringify(obj) + "\n"));
    return await this.#readLine();
  }

  async #readLine(): Promise<Record<string, unknown>> {
    while (!this.#buf.includes("\n")) {
      const chunk = new Uint8Array(4096);
      const n = await this.#conn.read(chunk);
      if (n === null) throw new Error("gateway closed the connection");
      this.#buf += this.#dec.decode(chunk.subarray(0, n));
    }
    const i = this.#buf.indexOf("\n");
    const line = this.#buf.slice(0, i);
    this.#buf = this.#buf.slice(i + 1);
    return JSON.parse(line);
  }

  close() {
    this.#conn.close();
  }
}

if (import.meta.main) {
  const path = Deno.args[0] ?? "./squeuelite.sock";
  const db = await Squeue.connect(path, "demo-deno");
  console.log("health:", await db.health());
  console.log(
    "write :",
    await db.execute(
      "INSERT INTO events(agent_id, kind) VALUES (?, ?)",
      ["demo-deno", "started"],
      { idempotencyKey: "demo:1" },
    ),
  );
  console.log("stats :", await db.stats());
  db.close();
}
